File: Pod/Classes/VISPERApplication.m
//
//  VISPERApplication.m
//  Pods
//
//  Created by Bartel on 20.08.15.
//
//

#import "VISPERApplication.h"
#import "VISPERWireframe.h"
#import "VISPERPushRoutingPresenter.h"
#import "VISPERModalRoutingPresenter.h"
#import "VISPERRootVCRoutingPresenter.h"
#import "VISPERReplaceTopVCRoutingPresenter.h"
#import "VISPERShowRoutingPresenter.h"
@import VISPER_CommandBus;

@interface VISPERApplication()
@property(nonatomic,strong)UINavigationController *navigationController;
@property(nonatomic,strong)NSObject<IVISPERWireframe> *wireframe;
@property(nonatomic,strong)VISPERCommandBus *commandBus;
@end

@implementation VISPERApplication

-(instancetype)init{
    return [self initWithNavigationController:nil
                                    wireframe:nil
                                   commandBus:nil];
}

-(instancetype)initWithNavigationController:(UINavigationController*)controller{
    return [self initWithNavigationController:controller wireframe:nil commandBus:nil];
}

-(instancetype)initWithNavigationController:(UINavigationController*)controller
                                  wireframe:(NSObject<IVISPERWireframe>*)wireframe{
    return [self initWithNavigationController:controller
                                    wireframe:wireframe
                                   commandBus:nil];
}

-(instancetype)initWithNavigationController:(UINavigationController*)controller
                                  wireframe:(NSObject<IVISPERWireframe>*)wireframe
                                 commandBus:(VISPERCommandBus*)commandBus{
    self = [super init];
    if(self){
        
        if(!wireframe){
            wireframe = [[VISPERWireframe alloc] init];
        }
        
        if(!commandBus){
            commandBus = [[VISPERCommandBus alloc] init];
        }
        
        if(!controller){
            controller = [[UINavigationController alloc] init];
        }
        
        self->_navigationController = controller;
        self->_wireframe = wireframe;
        self->_commandBus = commandBus;
        
        [self addRoutingPresenter:[[VISPERPushRoutingPresenter alloc] initWithNavigationController:self.navigationController]
                     withPriority:0];
        [self addRoutingPresenter:[[VISPERModalRoutingPresenter alloc] initWithNavigationController:self.navigationController]
                     withPriority:0];
        [self addRoutingPresenter:[[VISPERRootVCRoutingPresenter alloc] initWithNavigationController:self.navigationController]
                     withPriority:0];
        [self addRoutingPresenter:[[VISPERReplaceTopVCRoutingPresenter alloc] initWithNavigationController:self.navigationController]
                     withPriority:0];
        [self addRoutingPresenter:[[VISPERShowRoutingPresenter alloc] initWithNavigationController:self.navigationController]
                     withPriority:0];
        
        [self setNavigationController:controller];
    }
    return self;


}

-(UIViewController*)rootViewController{
    return self.navigationController;
}

-(void)addRoutingPresenter:(NSObject<IVISPERRoutingPresenter> *)routingPresenter
              withPriority:(NSInteger)priority{
    [self.wireframe addRoutingPresenter:routingPresenter withPriority:priority];
}

-(void)addFeature:(NSObject<IVISPERFeature> *)feature{
    
   if([feature respondsToSelector:@selector(bootstrapWireframe:commandBus:)]){
        [feature bootstrapWireframe:self.wireframe
                         commandBus:self.commandBus];
    }else{
        @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                       reason:@"a feature has to implement one of the following methods: bootstrapWireframe:commandBus: or bootstrapWireframe:interactor:"
                                     userInfo:nil];
    }
    
    
    if([feature respondsToSelector:@selector(routePatterns)]){
        NSArray *routePatterns = [feature routePatterns];
        
        for (NSString *routePattern in routePatterns) {
            [self.wireframe addRoute:routePattern];
        }
    }
}

-(void)addCommandHandler:(id)handler{
    [self.commandBus addHandler:handler];
}

-(BOOL)canRouteURL:(NSURL *)URL
    withParameters:(NSDictionary *)parameters{
    return [self.wireframe canRouteURL:URL withParameters:parameters];
}

-(BOOL)routeURL:(NSURL *)URL
 withParameters:(NSDictionary *)parameters
        options:(NSObject<IVISPERRoutingOption> *)options{
    return [self.wireframe routeURL:URL
                     withParameters:parameters
                            options:options];
}

-(UIViewController*)controllerForURL:(NSURL *)URL
                      withParameters:(NSDictionary *)parameters{
    return [self.wireframe controllerForURL:URL withParameters:parameters];
}


-(void)setNavigationController:(UINavigationController *)navigationController{
    self->_navigationController = navigationController;
    
    NSArray *routingPresenters = self.wireframe.routingPresenters;
    
    for(NSObject <IVISPERRoutingPresenter> *routingPresenter in routingPresenters){
        if([routingPresenter respondsToSelector:@selector(setNavigationController:)]){
            [routingPresenter performSelector:@selector(setNavigationController:) withObject:navigationController];
        }
    }
    
    [self.wireframe setCurrentViewController:self.navigationController.topViewController];
}

@end

File: VISPER/Classes/Deprecated/VISPERApplication.swift
//
//  VISPERApplication.swift
//  VISPER
//
//  Created by bartel on 28.12.17.
//

import Foundation
import VISPER_Core
import VISPER_CommandBus
import VISPER_Wireframe
import VISPER_Objc
import VISPER_Swift
import VISPER_Redux

open class VISPERApplication: NSObject,IVISPERApplication {
    
    var _navigationController: UINavigationController
    var _wireframe: IVISPERWireframe
    var _commandBus: VISPERCommandBus
    
    let application: AnyApplication<NSMutableDictionary>
    
    public override convenience init() {
        let controller = UINavigationController()
        self.init(navigationController: controller)
    }
    
    public convenience init!(navigationController controller: UINavigationController!) {
        
        let wireframe = DefaultWireframe()
        let wireframeObjc = WireframeObjc(wireframe: wireframe)
        let visperWireframe = VISPERWireframe(wireframe: wireframeObjc)
        
        self.init(navigationController: controller, wireframe: visperWireframe)
    }
    
    public convenience init!(navigationController controller: UINavigationController!, wireframe: VISPERWireframe!) {
        let commandBus = VISPERCommandBus()
        self.init(navigationController: controller, wireframe: wireframe, commandBus: commandBus)
    }
    
    public init!(navigationController controller: UINavigationController!,
                                       wireframe: VISPERWireframe!,
                                      commandBus: VISPERCommandBus!) {
        self._navigationController = controller
        self._wireframe = wireframe
        self._commandBus = commandBus
        
        //VISPERWireframe.addRoutingOptionConverter(converter: DefaultRoutingOptionConverter())
        
        let applicationFactory = ApplicationFactory<NSMutableDictionary>()
        let application = applicationFactory.makeApplication(initialState: NSMutableDictionary(),
                                                               appReducer: { (provider, action, state) -> NSMutableDictionary in
                                                return provider.reduce(action: action, state: state)
                                           },
                                           wireframe: wireframe.wireframe.wireframe)
        self.application = application
        
        let featureObserver = DeprecatedVISPERFeatureObserver<NSMutableDictionary>(wireframe: wireframe.wireframe.wireframe, commandBus: commandBus)
        self.application.add(featureObserver: featureObserver)
        
        self.application.add(controllerToNavigate: controller)
        
    }
    
    open func rootViewController() -> UIViewController! {
        return self._navigationController
    }
    
    open func wireframe() -> IVISPERWireframe! {
        return self._wireframe
    }
    
    open func commandBus() -> VISPERCommandBus! {
        return self._commandBus
    }
    
    open func navigationController() -> UINavigationController! {
        return self._navigationController
    }
    
    open func setNavigationController(_ navigationController: UINavigationController!) {
        self._navigationController = navigationController
        self.application.add(controllerToNavigate: navigationController)
    }
    
    open func add(_ feature: IVISPERFeature!) {
        let wrapper = DeprecatedVISPERFeatureWrapper(visperFeature: feature)
        do {
            try self.application.add(feature: wrapper)
        } catch let error {
            print("Error: \(error)")
        }
    }
    
    open func addCommandHandler(_ handler: Any!) {
        self.commandBus().addHandler(handler as AnyObject)
    }
    
    open func add(_ routingPresenter: IVISPERRoutingPresenter!, withPriority priority: Int) {
        self.wireframe().add(routingPresenter, withPriority: priority)
    }
    
    @discardableResult
    open func routeURL(_ URL: URL!, withParameters parameters: [AnyHashable : Any]!, options: IVISPERRoutingOption!) -> Bool {
        return self.wireframe().routeURL(URL,
                                         withParameters: parameters,
                                         options: options)
    }
    
    open func canRouteURL(_ URL: URL!, withParameters parameters: [AnyHashable : Any]!) -> Bool {
        return self.wireframe().canRouteURL(URL, withParameters: parameters)
    }
    
    open func controller(for URL: URL!, withParameters parameters: [AnyHashable : Any]!) -> UIViewController! {
        return self.wireframe().controller(for: URL, withParameters: parameters)
    }
    
    
}
